feat(selectEditors): add select grid editors

jmzagorski opened the pull request:
This pull request brings two new editors and one small change and should close https://github.com/ghiscoding/aurelia-slickgrid/issues/18

**Purpose**
Select editors are very common with data to ensure data integrity and we already provide these are header filters.

**Features**
1. SelectListEditor - this editor uses most of the logic from the `SelectListFilter`. If the developer does not use `multiple-select` library, this editor falls back to a bootstrap `form-control`. For this to render property in the grid a new css rules was added.
2. MultipleSelectListEditor - this editor uses most of the logic from the `MultipleSelectListFilter`. If the developer does not use `multiple-select` library, this editor falls back to a bootstrap `form-control`
3. CollectionFormatter - the formatter will lookup the `label` value in the `filter.collection` based on the value passed into the formatter. If it is an array, it will use the `arrayToCsvFormatter`
4. arraysEqual- a utility function to compare all values in an array for the `MultipleSelectListEditor`
5. Array.prototype.findOrDefault - An extension function to make code look cleaner especially when finding complex objects. If `find` returns undefined the consumer of this function can still safely try and access the property since by default the returned value is an empty object.

**Changes**
1. The arrayToCsvFormatter was changed to a span with a title in cases where the csv string is too long to show all values

**Tests**
I didn't start creating unit tests yet because `aurelia-pal-nodejs` needs to be updated because of this fix https://github.com/aurelia/pal-nodejs/issues/24 and another error i need to research about imports. Therefore, I manually tested with example 3 by changing the editor on the effort-driven column and adding/removing the multiple-select lib. 

**Comments**
1. (Future) We could probably improve the logic `this.labelName = (this.columnDef.filter.customStructure) ? this.columnDef.filter.customStructure.label : 'label';` since it is referenced in 3 files now. We could have a default `columnDef.filter` object that ensures the `customStructure` is always there.
2. (Future) We could refactor the logic to create the select DOM element since it is the same in the `Filter` and `Editor` or just wait to see if we can/want to suppose templates

- [ ] Update WIKI

ghiscoding commented:
Could you please modify 1 of the example, possibly Example 3, to include that editor? 
Changing the "% complete" with that editor could be a good demo.

jmzagorski commented:
added the `SingleSelectEditor` to the `% complete` and `MultipleSelectEditor` to a new `Prerequisites` column (it is the only new column I could think of that would have multiple values).

From the looks of it, I do not think the `setValues` needs to be in these editors after something changes in the grid since the editor is only active when a user interacts with the cell. You should be able to move a column and the editor should work the same. I have no experience yet with presets so you might want to test that example.

ghiscoding commented:
I added the `setValues` (only in Angular for now) because if I move a column (drag and change column position), the filter loses it's value because SlickGrid re-renders after moving a column. Have you tried changing a columns position while the editor is open? if the editor closes then no problem I guess. I'm in the middle of implementing the "Grid State & Grid Presets", so I can't really test it right now.

Actually thinking about it while writing, I think the editor goes away while you drag a column since it goes away onBlur, is that right? if so, no need for that `setValues`. The `setValues` is coming in my current feature.

jmzagorski commented:
It looks like the grid commits the change and closes the editor when moving a column while an editor is opened except for the new `MultipleSelectEditor` i just implemented. It behaves just like the other editors, but the select dialog box stays open.

ghiscoding commented:
I'm practically done with my PR #23 but I'd like to get this PR here in place before pushing a new release. I have left some comments which we can discuss

jmzagorski commented:
which comments did you want to discuss? Sorry I might have missed some

ghiscoding commented:
see the review comments on top

jmzagorski commented:
sorry I am looking for these "review comments" at the top and can't find them. I am probably looking in the wrong place and they are probably right in front of me somewhere. Once I find where those are I can review those tomorrow.

ghiscoding commented:
oh gosh I feel embarrassed, I thought it was live review, I didn't know that I had to send the review once it's completed. feel dump oopsy

jmzagorski commented on aurelia-slickgrid/src/aurelia-slickgrid/editors/multipleSelectEditor.ts:
```diff
@@ -0,0 +1,180 @@
+
+  init() {
+    if (!this.args) {
+      throw new Error('[Aurelia-SlickGrid] A filter must always have an "init()" ' +
```
I can change this to one line. I have a max line length of 80 in my editor. Do you have this setting in a file, if so what is it?

ghiscoding replied:
I don't have this setting and I don't follow it either. I understand it's for readability and I try to split into multiple lines when it's long but it's my own judgement

jmzagorski commented on aurelia-slickgrid/src/aurelia-slickgrid/editors/singleSelectEditor.ts:
```diff
@@ -0,0 +1,176 @@
+        '{ filter: type: FilterType.multipleSelect, collection: [{ value: true, label: \'True\' }, ' +
+        '{ value: false, label: \'False\'}] }');
+    }
+    this.optionCollection = this.columnDef.filter.collection || [];
```
I can change it to reference `params.collection`. That should be no problem.

So do you want me to change the `optionCollection` property name in the editor to be named `collection`? `optionCollection` should be a private property in the editor since it is just a reference to `collection`, and to be honest I cannot remember why I called the property `optionCollection`

jmzagorski replied:
What about accessing the `customStructure`? I use this in both the editors and the `collectionFormatter`? Should that be `params.customStructure` too?

jmzagorski replied:
And the reason I chose to use the `filter` object is because i assume that most of the time the `filter.collection` and `filter.customStructure` would be the same for the editors `collection` and `customStructure`

ghiscoding replied:
To make it to work in both Editor and Filter, we will have to review all the Filters, for example the MultipleFilter should use this instead 
```diff
-if (!this.columnDef || !this.columnDef.filter || !this.columnDef.filter.collection) {
+if (!this.columnDef || !this.columnDef.filter || !this.columnDef.filter.collection || this.columnDef.params.collection) {
  throw new Error(`[Aurelia-SlickGrid] You need to pass a "collection" for the MultipleSelect Filter to work correctly. Also each option should include a value/label pair (or value/labelKey when using Locale). For example:: { filter: type: FilterType.multipleSelect, collection: [{ value: true, label: 'True' }, { value: false, label: 'False'}] }`);
}

-const optionCollection = this.columnDef.filter.collection || [];
+const optionCollection = this.columnDef.filter.collection || this.columnDef.params.collection || [];
```

You can keep the `optionCollection` within the editor if you want but the one in the interface should really be `collection` to align with the already implemented Filters. As mentioned in my previous post, we can revisit that in `2.x` 

I didn't think about the `customStructure` but yeah it also makes sense to put it in the `params` as well. I understand the `filter` principle but still I would rather make a perfect cut between the 2 and not see anything about `filter` within an `editor`, it might confuse some users to do that, however `params` is generic enough not to confuse them.

jmzagorski replied:
`params` should be fine, the only time you might run into problems if there is more than one `collection` or `customStructure`. Not sure if that will ever happen in one column though?

Okay i see where i took `optionCollection` from now. You use it as a private `const` and I just promoted it to a property on the `this` variable within the editor so I could use it in other methods within the class. I really should make all those variables that are not part of the `Editor` interface private, for example
```typescript
  /**
   * The options label/value object to use in the select list
   */
  private _collection: SelectOption[] = [];
  /**
   * The property name for values in the optionCollection
   */
  private _valueName: string;
  /**
   * The property name for labels in the optionCollection
   */
  private _labelName: string;
```

ghiscoding replied:
oh ok, if you think it should be changed in Filter as well, I'm ok with that. Thanks for the comment

jmzagorski replied:
we don't have to do that right now because it might be better to change them all later in one commit **if** you want them all explicitly private. I will just change the `optionCollection` name

ghiscoding replied:
sure, that is more then ok too.

ghiscoding replied:
oh I guess that would be another thing that we can put inside the params and make the Editor and Filter share it, when provided

jmzagorski replied:
i can see `elementOptions` because it is for the actual dom element, but do we need to specify it is for the collection element? `collectionOptions`?

ghiscoding replied:
I think it's more for the `multiple-select` DOM element, so I would go more with `elementOptions` or `domElementOptions` but it's prob too long

jmzagorski replied:
I can go with `elementOptions`. The only downside I can see with this approach is if there were ever other elements in the column, then they would potentially have to share this `elementOptions`, but this might be how you want it designed since most DOM elements wont share many properties.

It will be a bit before I update this PR, but I will have it to you later this evening.

jmzagorski commented:
I'll try to get in the revisions at some point today or later tonight. I will not be in the office today and I will have limited access to my computer.

ghiscoding commented:
Sure whenever is best for you, no rush. However if you can't make it by tomorrow, then let me know and I'll cut a release tonight instead. Thanks

jmzagorski commented:
Okay I think I changed everything, but let me know if I missed something. I will be unavailable for the next few hours.

jmzagorski commented:
One more thing I'll need your opinion on that i just realized. The editors assume that the data item is a foreign key reference and not a complex object with a `label` and `value` property, see example 3 Where I set the data for those properties.

Is this a bad assumption and should I support a list of values and a list of complex objects with a `key` and `value`? Hopefully that makes sense.

ghiscoding commented on aurelia-slickgrid/src/aurelia-slickgrid/editors/singleSelectEditor.ts:
```diff
@@ -122,23 +121,22 @@ export class SingleSelectEditor implements Editor {
-          'Locale) is required to populate the Select list, for example:: { filter: ' +
-          'type: FilterType.multipleSelect, collection: [ { value: \'1\', label: \'One\' } ])');
+          'Locale) is required to populate the Select list, for example: ' +
+          '{ collection: [ { value: \'1\', label: \'One\' } ])');
```
You added an open curly brace in your throw error, though I think it's missing an ending curly brace.

ghiscoding commented on aurelia-slickgrid/src/aurelia-slickgrid/editors/multipleSelectEditor.ts:
```diff
@@ -0,0 +1,178 @@
+      container: 'body',
+      filter: false,
+      maxHeight: 200,
+      width: '100%',
```
you should not fix this to 100%, it makes the option "autoDropWidth" unusable. The "autoDropWidth" is a new option that I added which will resize the drop menu with the same size as the input select element.

ghiscoding commented:
Now that I am testing it locally, some observations
- intellisense doesn't work on `params`, since yeah it's of type `any`. I'm not totally in love with that anymore, however to kinda help, we can pass an extra type to `params: ExtraOptions` where `ExtraOptions` would be a new interface that includes `elementOptions: SelectOption` and `collection: any[]` but if we do that, then we might as well add a new Column property.
  - my bad, I should have thought about the intellisense much earlier. If we do add a new Column property, I would like to make it a generic to include both Editor/Filter and possibly other piece of code that might use a `collection`.
- you should not fix this to 100%, it makes the option "autoDropWidth" unusable. The "autoDropWidth" is a new option that I added which will resize the drop menu with the same size as the input select element. 
  - 
- it would also be better if you use the same defaults as the [multipleSelectFilter](https://github.com/ghiscoding/aurelia-slickgrid/blob/master/aurelia-slickgrid/src/aurelia-slickgrid/filters/multipleSelectFilter.ts#L38), the translation key are of particular interest because if they are omitted then the 3rd party lib (in this case `multiple-select.js`) will use it's own titles (only English). So please provide them.

jmzagorski commented:
Thank you for the feedback. I will look this over and answer your questions a bit later when I am near my computer. Not sure how late I am able to work tonight so you can cut a release without this if you like.

ghiscoding commented:
I can wait another day, no hurry. 
No one complained about the other bug yet so it's ok to wait another day.

jmzagorski commented:
> question. Do you want to keep params for now and come back with another property later? Or do you want to create a new Column property now?

If you are not opposed, lets keep `params` since there are a lot of changes in this PR already. Then we can have another PR focused on just updating the API for the `Editor` and `Filter`

> it would also be better if you use the same defaults as the multipleSelectFilter,

I think this might be a larger issue because we do not have control over creating an editor to inject the `i18n` library. The only way for dependency injection to work in an `Editor` is for the user to inject the editor into their view model using `Factory.Of` or (which is better in my opinion) we wrap their `columnDef.editor` property in `Factory.of` in the `AureliaSlickGridCustomElement`or somewhere in the library.

ghiscoding commented:
Yes sure, we can always create more PR. I am not opposed to that at all, I always prefer to ask :)

You can also use `params` to pass the `i18n` this is actually the preferred way of doing it in my [Wiki - localization](/ghiscoding/aurelia-slickgrid/wiki/Localization#passing-i18n-in-the-grid-options) with Formatters. That is what I do for Formatter and it will work perfectly with Editor as well, you just use the instance without DI. The good old javascript way. Welcome back to the javascript world lol ;)

ghiscoding commented:
Since I have that in mind, I wanted to tell you yesterday that the [dateEditor](/ghiscoding/aurelia-slickgrid/blob/master/aurelia-slickgrid/src/aurelia-slickgrid/editors/dateEditor.ts#L47) does use the `i18n` in it. Which is like I told you, I pull the `i18n` from the `params.i18n`, it has to be written that way and it is in my [Wiki docs](/ghiscoding/aurelia-slickgrid/wiki/Localization#passing-i18n-in-the-grid-options).

jmzagorski commented:
Okay let's see if i fixed all your notes:

1. Editor will use `params` object for now until the new `params` interface is defined in v2.0 
2. I set the width of the `multiple-select` to the values you provided and it looks good
3. Added i18n translation in both editors so the behavior is similar to the filter. 
  - [ ] Opened an issue to further discuss pros/cons of allowing DI in editors since the library uses `i18n` for some editors. We therefore would not be dependent on the user to pass this.
4. Per your suggestion I added the `setTimeout` and `open` method and it behaves a lot nicer since the select lists opens immediately. However, after doing some research, this does not seem to be easy with the fallback HTML single select element so I left it as is. Finally, I limited the drop down to Task0 - Task10 because for the demo.

ghiscoding commented on aurelia-slickgrid/src/aurelia-slickgrid/editors/multipleSelectEditor.ts:
```diff
@@ -39,18 +40,31 @@ export class MultipleSelectEditor implements Editor {
+  private _i18n: I18N;
 
   constructor(private args: any) {
+    this._i18n = this.args.column.params.i18n;
```
this will probably fail at Build, `args` could possibly be undefined here

ghiscoding commented on aurelia-slickgrid/src/aurelia-slickgrid/editors/singleSelectEditor.ts:
```diff
@@ -38,13 +39,20 @@ export class SingleSelectEditor implements Editor {
+  private _i18n: I18N;
 
   constructor(private args: any) {
+    this._i18n = this.args.column.params.i18n;
```
same as before.. this will probably fail at Build, `args` could possibly be undefined here

ghiscoding commented:
Awesome this looks really nice, I tried it over lunch and yes you did addressed all the things I said.
The last thing would be to see if errors comes up in Build Lib, I'm expecting that you might have some errors with `possibly undefined`

jmzagorski commented:
I ran the build and I did not receive any errors. Did you see errors on your end?

For the offset, do I have to somehow dynamically set that value if it is the last column or always setting that value the way I did it won't hurt anything?

ghiscoding commented:
I didn't see any errors, I assumed the Build would throw an error. If it's ok then fine
